### PR TITLE
Describe paradigms used by Flutter

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
This is a continuation of the PR in https://github.com/flutter/website/pull/651.

File: faq.md [hunk 1]
```diff
@@ -500,6 +500,82 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
 or **Release Mode**.
 
+### What programming paradigm does Flutter's framework use?
+
+Flutter is a multi-paradigm programming environment. Many programming
+techniques developed over the past few decades are used in Flutter. We
+use each one where we believe the strengths of the technique make it
+particularly well-suited. In no particular order:
+
+* Composition: The primary paradigm used by Flutter is that of using small objects with narrow scopes of behavior,
```
Comment (sethladd) [inline faq.md]: when possible, please break lines at <= 80 cols. Makes reviews easier.

File: faq.md [hunk 2]
```diff
@@ -500,6 +500,82 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+coordinate system, and then we have a subclass
+([RenderBox](https://docs.flutter.io/flutter/rendering/RenderBox-class.html)) that introduces the opinion that the
+geometry should be based on the Cartesian coordinate system (x/width and y/height).
+
+* Prototype-based object-orientated programming: The
+[ScrollPhysics](https://docs.flutter.io/flutter/widgets/ScrollPhysics-class.html) class chains instances to compose the
+physics that apply to scrolling dynamically at runtime. This lets the system compose, for example, paging physics with
+platform-specific physics, without the platform having to be selected at compile time.
+
+* Imperative programming: Straight forward imperative programming, usually paired with state encapsulated within an
```
Comment (sethladd) [inline faq.md]: I think it's "straightfoward"

**Comment by Hixie** [comment]
Thanks @sethladd 

@wmleler definitely still want your review as well since you were the main person with feedback on the original version of this PR. (Flutter convention is to make sure everyone who has an opinion is on board, it's considered very much out-of-line to check something in without the LGTM from someone who has raised objections.)

**Comment by Hixie** [comment]
I've added examples to each subsection. @wmleler says he is busy until Oct 3 but will review then.

**Comment by mit-mit** [comment]
cc @wmleler now that he is back in the office

**Comment by wmleler** [comment]
I'm back! Will do this ASAP. Thanks for being patient.

File: faq.md [hunk 3]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+an [AnimatedDefaultTextStyle](https://docs.flutter.io/flutter/widgets/AnimatedDefaultTextStyle-class.html),
+and a [ConstrainedBox](https://docs.flutter.io/flutter/widgets/ConstrainedBox-class.html).
+The [InkWell](https://docs.flutter.io/flutter/material/InkWell-class.html)
+is built using a [GestureDetector](https://docs.flutter.io/flutter/widgets/GestureDetector-class.html).
+The [Material](https://docs.flutter.io/flutter/material/Material-class.html)
+is built using an [AnimatedDefaultTextStyle](https://docs.flutter.io/flutter/widgets/AnimatedDefaultTextStyle-class.html),
+a [NotificationListener](https://docs.flutter.io/flutter/widgets/NotificationListener-class.html),
+and an [AnimatedPhysicalModel](https://docs.flutter.io/flutter/widgets/AnimatedPhysicalModel-class.html).
+And so on. It's widgets all the way down.
+
```
Comment (wmleler) [inline faq.md]: When you say "Entire applications can be built with only StatelessWidget" you should probably mention the limitations. Like "Entire applications, such as those with no user input widgets, can be built with only StatelessWidget"

File: faq.md [hunk 4]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+* Functional programming: Entire applications can be built with only
+[StatelessWidget](https://docs.flutter.io/flutter/widgets/StatelessWidget-class.html)s,
+which are essentially functions that describe how arguments map to other
+functions, bottoming out in primitives that compute layouts or paint graphics.
+For example, the [Icon](https://docs.flutter.io/flutter/widgets/Icon-class.html)
+widget is essentially just a function that maps its arguments
```
Comment (wmleler) [inline faq.md]: remove "just"

File: faq.md [hunk 5]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+paired with state encapsulated within an object, is used where it provides the
+most intuitive solution. For example, tests are written in an imperative style,
+first describing the situation under test, then listing the invariants that the
+test must match, then advancing the clock or inserting events as necessary for
+the test.
+
+* Reactive programming: The widget and element trees are sometimes described as
+reactive, because new inputs provided in a widget's constructor are immediately
+propagated as changes to lower-level widgets by the widget's build method, and
+changes made in the lower widgets (e.g. in response to user input) propagate
```
Comment (wmleler) [inline faq.md]: Add comma to make it e.g.,

File: faq.md [hunk 6]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+selected at compile time.
+
+* Imperative programming: Straightforward imperative programming, usually
+paired with state encapsulated within an object, is used where it provides the
+most intuitive solution. For example, tests are written in an imperative style,
+first describing the situation under test, then listing the invariants that the
+test must match, then advancing the clock or inserting events as necessary for
+the test.
+
+* Reactive programming: The widget and element trees are sometimes described as
```
Comment (wmleler) [inline faq.md]: Maybe change "sometimes" to "popularly"?

File: faq.md [hunk 7]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+reactive, because new inputs provided in a widget's constructor are immediately
+propagated as changes to lower-level widgets by the widget's build method, and
+changes made in the lower widgets (e.g. in response to user input) propagate
+back up the tree via event handlers. Aspects of both functional-reactive and
+imperative-reactive are present in the framework, depending on the needs of the
+widgets. For example, the material
+[Divider](https://docs.flutter.io/flutter/material/Divider-class.html) class has
+a build method that consists of just an expression describing how this widget
+reacts to changes in its configuration (a functional reactive widget), while the
```
Comment (wmleler) [inline faq.md]: Move the parenthetical remark to right after the name "Divider class".

File: faq.md [hunk 8]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+[Chip](https://docs.flutter.io/flutter/material/Chip-class.html) class has a
+build method that constructs a list of children over several statements,
+describing how this widget reacts to changes in its configuration (an imperative
+reactive widget).
```
Comment (wmleler) [inline faq.md]: Again, move the parenthetical remark to right after "Chip class"

File: faq.md [hunk 9]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+
+* Declarative programming: The build methods of widgets are often a single
+expression with multiple levels of nested constructors, written using a strictly
+declarative subset of Dart (such nested expressions could be mechanically
```
Comment (wmleler) [inline faq.md]: "(i.e., such nested expressions..."

File: faq.md [hunk 10]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+transformed to or from any suitably expressive markup language). For example,
+the
+[UserAccountsDrawerHeader](https://docs.flutter.io/flutter/material/UserAccountsDrawerHeader-class.html)
+widget has a long build method (20+ lines), consisting of a single nested
+expression. This can also be combined with the imperative style to build UIs
+that would be harder to describe in a pure-declarative approach.
+
+* Generic programming: Types can be used to help catch programming errors early.
```
Comment (wmleler) [inline faq.md]: delete "help"

File: faq.md [hunk 11]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+The Flutter framework uses generic programming to help in this regard. For
+example, the [State](https://docs.flutter.io/flutter/widgets/State-class.html)
+class is parameterized in terms of the type of its associated widget, so that
+the Dart analyzer can catch mismatches of states and widgets. Similarly, the
+[GlobalKey](https://docs.flutter.io/flutter/widgets/GlobalKey-class.html) class
+takes a type parameter so that it can access a remote widget's state in a
+type-safe manner (with runtime checking), the
```
Comment (wmleler) [inline faq.md]: Change "with" to "using"

File: faq.md [hunk 12]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+[Route](https://docs.flutter.io/flutter/widgets/Route-class.html) interface is
+parameterized with the type that it is expected to use when
+[popped](https://docs.flutter.io/flutter/widgets/Navigator/pop.html), and
+collections such as
+[List](https://docs.flutter.io/flutter/dart-core/List-class.html)s,
+[Map](https://docs.flutter.io/flutter/dart-core/Map-class.html)s, and
+[Set](https://docs.flutter.io/flutter/dart-core/Set-class.html)s are all
+parameterized so that mismatched elements can be caught early either by the
```
Comment (wmleler) [inline faq.md]: "can be caught during debugging by either the analyzer or at runtime."

File: faq.md [hunk 13]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+analyzer or at runtime during debugging.
+
+* Concurrent programming: Flutter makes heavy use of
```
Comment (wmleler) [inline faq.md]: Wouldn't it just be better to call this section "Futures" instead of "Concurrent programming"?

File: faq.md [hunk 14]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+[Future](https://docs.flutter.io/flutter/dart-async/Future-class.html)s and
+other asynchronous APIs. For example, the animation system reports when an
+animation is finished by completing a future. The image loading system similarly
+uses futures to report when a load is complete.
+
+* Constraint programming: The layout system in Flutter uses a weak form of
+constraint programming to determine the geometry of a scene. Constraints (e.g.
```
Comment (wmleler) [inline faq.md]: "e.g." must always be followed by a comma (same thing for i.e.)

File: faq.md [hunk 15]
```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
...
+for cartesian boxes, a minimum and maximum width and a minimum and maximum
+height) are passed from parent to child, and the child selects a resulting
+geometry (e.g. for cartesian boxes, a size, specifically a width and a height)
+that fulfills those constraints. By using this technique, Flutter can usually
+lay out a whole scene with a single pass.
```
Comment (wmleler) [inline faq.md]: Change to "an entire scene"

File: faq.md (hunk 3 above)
Comment (Hixie) [reply faq.md]: Actually the very first application to be written with the runApp API was entirely stateless but still had user input (it just called runApp again when you provided input :-) ).

Added "`(Such applications can't easily have state, so are typically non-interactive.)`".

File: faq.md (hunk 4 above)
Comment (Hixie) [reply faq.md]: done

File: faq.md (hunk 6 above)
Comment (Hixie) [reply faq.md]: Not sure I can justify "popularly".

File: faq.md (hunk 5 above)
Comment (Hixie) [reply faq.md]: Done. Apparently this is an Americanism.

File: faq.md (hunk 7 above)
Comment (Hixie) [reply faq.md]: The parenthetical is trying to clarify "an expression describing how this widget reacts to changes in its configuration", not "Divider". I think putting the parenthetical after Divider would imply that some widgets aren't "functional reactive widgets".

File: faq.md (hunk 8 above)
Comment (Hixie) [reply faq.md]: same here, this is trying to distinguish "an expression describing how this widget reacts to changes in its configuration" (the earlier text) from "a build method that constructs a list of children over several statements, describing how this widget reacts to changes in its configuration".

File: faq.md (hunk 9 above)
Comment (Hixie) [reply faq.md]: Not sure the `i.e.` would add anything here.

File: faq.md (hunk 10 above)
Comment (Hixie) [reply faq.md]: Why? The types themselves don't catch any errors. It's the programmer who catches the errors with the help of the analyzer which itself is helped by the type declarations in generics.

File: faq.md (hunk 11 above)
Comment (Hixie) [reply faq.md]: done

File: faq.md (hunk 12 above)
Comment (Hixie) [reply faq.md]: changed to "either during analysis or at runtime during debugging". The analyzer isn't part of debugging.

File: faq.md (hunk 13 above)
Comment (Hixie) [reply faq.md]: The labels are those used by Wikipedia.

File: faq.md (hunk 14 above)
Comment (Hixie) [reply faq.md]: Done, but FYI this is only true of (some) American style guides.

File: faq.md (hunk 15 above)
Comment (Hixie) [reply faq.md]: done

**Comment by Hixie** [comment]
Updated.

File: faq.md (hunk 6 above)
Comment (wmleler) [reply faq.md]: I was using "popularly" in the sense of "technically sloppy".

File: faq.md (hunk 7 above)
Comment (wmleler) [reply faq.md]: The problem I was trying to solve was that you only told me which kind of example this is at the end of the example, which makes it hard to read. You should (somehow) indicate that the first example is the functional one, and the second one is the imperative one.

File: faq.md (hunk 9 above)
Comment (wmleler) [reply faq.md]: The "i.e.," is needed to make this easier to read.

File: faq.md (hunk 10 above)
Comment (wmleler) [reply faq.md]: If that is true, then why are you leaving out the subject (who/what is catching the errors)?

**Comment by wmleler** [comment]
a few minor comments, which you can deal with as you wish.

File: faq.md (hunk 6 above)
Comment (Hixie) [reply faq.md]: Why would it be sloppy? They are literally reactive in the sense described on Wikipedia, as far as I can tell.

File: faq.md (hunk 7 above)
Comment (Hixie) [reply faq.md]: Ah, good call. Fixed.

File: faq.md (hunk 9 above)
Comment (Hixie) [reply faq.md]: I made it a separate sentence.

File: faq.md (hunk 10 above)
Comment (Hixie) [reply faq.md]: Ah, I see. Added "developers".

**Comment by Hixie** [comment]
Updated. Will land on green.

Thanks for the thorough review!

**Comment by Hixie** [comment]
Travis failure appears to be bogus (it complains about a bad link but the link works fine for me). Going to retry. cc @sethladd

**Pull request merged**